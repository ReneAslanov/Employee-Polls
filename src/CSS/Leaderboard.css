.leaderboard-wrapper{
    width: 70vw;
    display: flex;
    flex-direction: column;
    margin: 2rem auto;
    border: 1px solid #e1e5eb;
    border-radius: 4px;
}

.leaderboard-head{
    display: flex;
    flex-direction: row;
}

.leaderboard-users{
    font-size: 1.2em;
    font-weight: 200;
    flex-basis: 0;
    flex-grow: 1;
    padding: 1em;
    border: 1px solid #e1e5eb;
}

.first-item{
    flex-grow: 3;
}

.leaderboard-items-wrapper{
    display: flex;
}

.leaderboard-item{
    font-size: 1.2em;
    font-weight: 200;
    flex-basis: 0;
    flex-grow: 1;
    padding: 1em;
    border: 1px solid #e1e5eb;
}

.leaderboard-user-name{
    flex-grow: 3;
}

.leaderboard-userName{
    font-size: 1em;
    font-weight: 200;
}

.leaderboard-userId{
    font-size: 1em;
    font-weight: 200;
}

.leaderboard-user-name{
    display: flex;
    flex-direction: row;
    gap: 1.2rem;
}

.leaderboard-userAvatar{
    width: 3.5rem;
    height: 3.5rem;
}