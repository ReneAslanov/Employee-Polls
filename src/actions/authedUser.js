export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT = "LOGOUT";
export const SET_POLL_OPTION = "SET_POLL_OPTION";

export function setAuthedUser(payload)
{
    return{
        type: SET_AUTHED_USER,
        id: payload.id,
        avatar: payload.avatarURL,
        answers: payload.answers,
        questions: payload.questions
    }
}

export function logoutUser()
{
    return{
        type: LOGOUT
    }
}

export function setPollOption({qid, answer})
{
    return{
        type: SET_POLL_OPTION,
        qid,
        answer
    }
}