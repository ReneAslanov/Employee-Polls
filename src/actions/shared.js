export const RECEIVE_USERS = "RECEIVE_USERS";
export const SET_POLL_ANSWER = "SET_POLL_ANSWER";
export const SET_QUESTION_ARRAY = "SET_QUESTION_ARRAY";

export function setUserAnswer({authedUser, qid, answer})
{
    return{
        type: SET_POLL_ANSWER,
        authedUser,
        qid,
        answer
    }
}

export function setUserQuestionArray({authedUser, qid})
{
  return{
    type: SET_QUESTION_ARRAY,
    authedUser,
    qid
  }
}

export function receiveUsers(payload)
{
    return{
        type: RECEIVE_USERS,
        users: payload
    }
}
