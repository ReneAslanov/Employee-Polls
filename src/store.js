import { legacy_createStore as createStore} from "redux";
import middleware from './middleware';
import reducer from "./reducers/index";

const store = createStore(reducer, middleware);

export default store;

