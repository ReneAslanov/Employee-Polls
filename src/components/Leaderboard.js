import { connect} from "react-redux";
import Nav from "./Nav";
import "../CSS/Leaderboard.css";
import LeaderboardCard from "./Leaderboard-Card";
import ErrorPage from "./ErrorPage";

function Leaderboard(props)
{
    if(props.authedUser === null)
    {
        return(
            <ErrorPage/>
        )
    }

    return(
        <div>
            <Nav />
            <div className="leaderboard-wrapper" data-testid="leaderboard-wrapper">
                <div className="leaderboard-head">
                    <div className="leaderboard-users first-item">Users</div>
                    <div className="leaderboard-users">Answered</div>
                    <div className="leaderboard-users">Created</div>
                </div>
                {
                    props.users.map(user => {
                        return <LeaderboardCard user={user} key={user.id}/>
                    })
                }
            </div>
        </div>
    )
}

function mapStateToProps({authedUser, users})
{
    if(authedUser === null)
    {
        return{
            authedUser: null
        }
    }

    return{
        users: Object.values(users).sort((a, b) => {
           return (Object.keys(b.answers).length + b.questions.length) - (Object.keys(a.answers).length + a.questions.length)
        })
    }

}

export default connect(mapStateToProps)(Leaderboard);