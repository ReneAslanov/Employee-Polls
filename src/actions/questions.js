import { _saveQuestionAnswer } from "../utils/_DATA.js"
import { SET_POLL_OPTION } from "./authedUser.js";

export const SET_QUESTIONS = "SET_QUESTIONS";
export const CREATE_QUESTION = "CREATE_QUESTION";

export function setQuestions(questions)
{
    return{
        type: SET_QUESTIONS,
        questions : Object.values(questions)
    }
}

export function createQuestion(question)
{
    return{
        type: CREATE_QUESTION,
        payload: question
    }
}

export function handleCreateQuestion(optionOne, optionTwo, author)
{

    return{
        optionOneText: optionOne,
        optionTwoText: optionTwo,
        author
    }
}

export function handleQuestionAnswer({authedUser, qid, answer})
{
    return (dispatch) => {
        return _saveQuestionAnswer({
            authedUser,
            qid,
            answer
        })
        .then(data => {
            return dispatch(setPollAnswer({authedUser, qid, answer}))
        })

    }
}

export function setPollAnswer({qid, answer, authedUser})
{
    return{
        type: SET_POLL_OPTION,
        qid,
        answer,
        authedUser
    }
}