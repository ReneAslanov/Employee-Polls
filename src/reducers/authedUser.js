import { SET_AUTHED_USER, SET_POLL_OPTION } from "../actions/authedUser";
import { LOGOUT } from "../actions/authedUser";
import { CREATE_QUESTION } from "../actions/questions";

export default function authedUser(state = null, action)
{
    switch(action.type)
    {
        case SET_AUTHED_USER:
            return{
               id: action.id,
               avatar: action.avatar,
               answers: action.answers,
               questions: action.questions
            };

        case LOGOUT:
            return null

        case CREATE_QUESTION:
            return{
                ...state,
                questions: state.questions.concat(action.payload.id)
            }

        case SET_POLL_OPTION:

            const {qid, answer} = action;
            return{
                ...state,
                answers: {
                    ...state.answers,
                    [qid]: answer
                }
            }

        default:
            return state;
    }
}