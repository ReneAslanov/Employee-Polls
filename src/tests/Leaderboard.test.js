import Leaderboard from "../components/Leaderboard";
import store from "../store";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { _getUsers } from "../utils/_DATA";
import { receiveUsers } from "../actions/shared";
import { setAuthedUser } from "../actions/authedUser";

describe("setAuthedUser", () => {
    it("should create an Object, which is then passed to the reducer", async() => {
        const testAuthedUser = {
            id: "sarahedo",
            avatarURL: "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU",
            answers: {
                "8xf0y6ziyjabvozdd253nd": 'optionOne',
                "6ni6ok3ym7mf1p33lnez": 'optionOne',
                "am8ehyc8byjqgar0jgpub9": 'optionTwo',
                "loxhs1bqm25b708cmbf3g": 'optionTwo'
              },
            questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
        }

        const testObject = setAuthedUser(testAuthedUser);
        expect(testObject).toBeInstanceOf(Object);
        expect(testObject.id).toBe("sarahedo");
        expect(testObject.questions.length).toEqual(2);
    })
})

describe("receiveUsers", () => {
    it("should return an object with all the users", async() => {
        const users = await _getUsers();
        const testUserArray = ["sarahedo", "tylermcginnis", "mtsamis", "zoshikanlu"];

        expect(users).toBeInstanceOf(Object);
        expect(Object.keys(users)).toEqual(testUserArray);
    })
})

describe("Leaderboard", () => {
    it("should render the Leaderboard component properly", async() => {
        const testAuthedUser = {
            id: "sarahedo",
            avatarURL: "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU",
            answers: {
                "8xf0y6ziyjabvozdd253nd": 'optionOne',
                "6ni6ok3ym7mf1p33lnez": 'optionOne',
                "am8ehyc8byjqgar0jgpub9": 'optionTwo',
                "loxhs1bqm25b708cmbf3g": 'optionTwo'
              },
            questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
        }

        await _getUsers().then(data => {
            store.dispatch(setAuthedUser(testAuthedUser));
            store.dispatch(receiveUsers(data));
            return true;
        });

         let view = render(
        <Provider store={store}>
            <BrowserRouter>
                <Leaderboard />
            </BrowserRouter>
        </Provider>);

        const wrapper = screen.getByTestId("leaderboard-wrapper");
        expect(view).toMatchSnapshot();
        expect(wrapper).toBeInTheDocument();
    })
})