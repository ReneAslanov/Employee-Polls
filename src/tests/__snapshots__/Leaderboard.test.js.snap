// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard should render the Leaderboard component properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="nav-wrapper"
        >
          <div
            class="nav-routes"
          >
            <ul
              class="nav-list"
            >
              <a
                href="/home"
              >
                Home
              </a>
              <a
                href="/leaderboard"
              >
                Leaderboard
              </a>
              <a
                href="/add"
              >
                New
              </a>
            </ul>
          </div>
          <div
            class="nav-user"
          >
            <ul
              class="nav-list"
            >
              <div
                class="nav-img-wrapper"
              >
                <img
                  alt="just an avatar"
                  class="nav-avatar"
                  src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU"
                />
              </div>
              <p
                class="nav-userName"
              >
                sarahedo
              </p>
              <a
                href="/"
              >
                Logout
              </a>
            </ul>
          </div>
        </div>
        <div
          class="leaderboard-wrapper"
          data-testid="leaderboard-wrapper"
        >
          <div
            class="leaderboard-head"
          >
            <div
              class="leaderboard-users first-item"
            >
              Users
            </div>
            <div
              class="leaderboard-users"
            >
              Answered
            </div>
            <div
              class="leaderboard-users"
            >
              Created
            </div>
          </div>
          <div
            class="leaderboard-items-wrapper"
          >
            <div
              class="leaderboard-user-name leaderboard-item"
            >
              <div
                class="leaderboard-user-wrapper"
              >
                <h3
                  class="leaderboard-userName"
                >
                  Sarah Edo
                </h3>
                <h4
                  class="leaderboard-userId"
                >
                  sarahedo
                </h4>
              </div>
              <div
                class="leaderboard-userAvatar-wrapper"
              >
                <img
                  alt="just an avatar"
                  class="leaderboard-userAvatar"
                  src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU"
                />
              </div>
            </div>
            <div
              class="leaderboard-user-answers leaderboard-item"
            >
              4
            </div>
            <div
              class="leaderboard-user-created leaderboard-item"
            >
              2
            </div>
          </div>
          <div
            class="leaderboard-items-wrapper"
          >
            <div
              class="leaderboard-user-name leaderboard-item"
            >
              <div
                class="leaderboard-user-wrapper"
              >
                <h3
                  class="leaderboard-userName"
                >
                  Mike Tsamis
                </h3>
                <h4
                  class="leaderboard-userId"
                >
                  mtsamis
                </h4>
              </div>
              <div
                class="leaderboard-userAvatar-wrapper"
              >
                <img
                  alt="just an avatar"
                  class="leaderboard-userAvatar"
                  src="http://cdn.onlinewebfonts.com/svg/img_81879.png"
                />
              </div>
            </div>
            <div
              class="leaderboard-user-answers leaderboard-item"
            >
              3
            </div>
            <div
              class="leaderboard-user-created leaderboard-item"
            >
              2
            </div>
          </div>
          <div
            class="leaderboard-items-wrapper"
          >
            <div
              class="leaderboard-user-name leaderboard-item"
            >
              <div
                class="leaderboard-user-wrapper"
              >
                <h3
                  class="leaderboard-userName"
                >
                  Tyler McGinnis
                </h3>
                <h4
                  class="leaderboard-userId"
                >
                  tylermcginnis
                </h4>
              </div>
              <div
                class="leaderboard-userAvatar-wrapper"
              >
                <img
                  alt="just an avatar"
                  class="leaderboard-userAvatar"
                  src="https://static.thenounproject.com/png/4035889-200.png"
                />
              </div>
            </div>
            <div
              class="leaderboard-user-answers leaderboard-item"
            >
              2
            </div>
            <div
              class="leaderboard-user-created leaderboard-item"
            >
              2
            </div>
          </div>
          <div
            class="leaderboard-items-wrapper"
          >
            <div
              class="leaderboard-user-name leaderboard-item"
            >
              <div
                class="leaderboard-user-wrapper"
              >
                <h3
                  class="leaderboard-userName"
                >
                  Zenobia Oshikanlu
                </h3>
                <h4
                  class="leaderboard-userId"
                >
                  zoshikanlu
                </h4>
              </div>
              <div
                class="leaderboard-userAvatar-wrapper"
              >
                <img
                  alt="just an avatar"
                  class="leaderboard-userAvatar"
                  src="https://www.seekpng.com/png/detail/514-5147366_default-avatar-comments-avatar.png"
                />
              </div>
            </div>
            <div
              class="leaderboard-user-answers leaderboard-item"
            >
              1
            </div>
            <div
              class="leaderboard-user-created leaderboard-item"
            >
              0
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="nav-wrapper"
      >
        <div
          class="nav-routes"
        >
          <ul
            class="nav-list"
          >
            <a
              href="/home"
            >
              Home
            </a>
            <a
              href="/leaderboard"
            >
              Leaderboard
            </a>
            <a
              href="/add"
            >
              New
            </a>
          </ul>
        </div>
        <div
          class="nav-user"
        >
          <ul
            class="nav-list"
          >
            <div
              class="nav-img-wrapper"
            >
              <img
                alt="just an avatar"
                class="nav-avatar"
                src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU"
              />
            </div>
            <p
              class="nav-userName"
            >
              sarahedo
            </p>
            <a
              href="/"
            >
              Logout
            </a>
          </ul>
        </div>
      </div>
      <div
        class="leaderboard-wrapper"
        data-testid="leaderboard-wrapper"
      >
        <div
          class="leaderboard-head"
        >
          <div
            class="leaderboard-users first-item"
          >
            Users
          </div>
          <div
            class="leaderboard-users"
          >
            Answered
          </div>
          <div
            class="leaderboard-users"
          >
            Created
          </div>
        </div>
        <div
          class="leaderboard-items-wrapper"
        >
          <div
            class="leaderboard-user-name leaderboard-item"
          >
            <div
              class="leaderboard-user-wrapper"
            >
              <h3
                class="leaderboard-userName"
              >
                Sarah Edo
              </h3>
              <h4
                class="leaderboard-userId"
              >
                sarahedo
              </h4>
            </div>
            <div
              class="leaderboard-userAvatar-wrapper"
            >
              <img
                alt="just an avatar"
                class="leaderboard-userAvatar"
                src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS3_juC_cG7iLmwAaK4Mxxy-OtJ52335hZfHhiS5eSHGp9a0C9OY67183F9TV4t01DNILs&usqp=CAU"
              />
            </div>
          </div>
          <div
            class="leaderboard-user-answers leaderboard-item"
          >
            4
          </div>
          <div
            class="leaderboard-user-created leaderboard-item"
          >
            2
          </div>
        </div>
        <div
          class="leaderboard-items-wrapper"
        >
          <div
            class="leaderboard-user-name leaderboard-item"
          >
            <div
              class="leaderboard-user-wrapper"
            >
              <h3
                class="leaderboard-userName"
              >
                Mike Tsamis
              </h3>
              <h4
                class="leaderboard-userId"
              >
                mtsamis
              </h4>
            </div>
            <div
              class="leaderboard-userAvatar-wrapper"
            >
              <img
                alt="just an avatar"
                class="leaderboard-userAvatar"
                src="http://cdn.onlinewebfonts.com/svg/img_81879.png"
              />
            </div>
          </div>
          <div
            class="leaderboard-user-answers leaderboard-item"
          >
            3
          </div>
          <div
            class="leaderboard-user-created leaderboard-item"
          >
            2
          </div>
        </div>
        <div
          class="leaderboard-items-wrapper"
        >
          <div
            class="leaderboard-user-name leaderboard-item"
          >
            <div
              class="leaderboard-user-wrapper"
            >
              <h3
                class="leaderboard-userName"
              >
                Tyler McGinnis
              </h3>
              <h4
                class="leaderboard-userId"
              >
                tylermcginnis
              </h4>
            </div>
            <div
              class="leaderboard-userAvatar-wrapper"
            >
              <img
                alt="just an avatar"
                class="leaderboard-userAvatar"
                src="https://static.thenounproject.com/png/4035889-200.png"
              />
            </div>
          </div>
          <div
            class="leaderboard-user-answers leaderboard-item"
          >
            2
          </div>
          <div
            class="leaderboard-user-created leaderboard-item"
          >
            2
          </div>
        </div>
        <div
          class="leaderboard-items-wrapper"
        >
          <div
            class="leaderboard-user-name leaderboard-item"
          >
            <div
              class="leaderboard-user-wrapper"
            >
              <h3
                class="leaderboard-userName"
              >
                Zenobia Oshikanlu
              </h3>
              <h4
                class="leaderboard-userId"
              >
                zoshikanlu
              </h4>
            </div>
            <div
              class="leaderboard-userAvatar-wrapper"
            >
              <img
                alt="just an avatar"
                class="leaderboard-userAvatar"
                src="https://www.seekpng.com/png/detail/514-5147366_default-avatar-comments-avatar.png"
              />
            </div>
          </div>
          <div
            class="leaderboard-user-answers leaderboard-item"
          >
            1
          </div>
          <div
            class="leaderboard-user-created leaderboard-item"
          >
            0
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
